File: NextcloudTalk/Chat/VoiceMessageTranscribeViewController.m
/**
 * SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#import <Speech/Speech.h>

#import "VoiceMessageTranscribeViewController.h"
#import "NCAppBranding.h"

@interface VoiceMessageTranscribeViewController () {
    UIActivityIndicatorView *_activityIndicator;
    NSURL *_audioFileUrl;
    NSArray *_supportedLocales;
}

@end

@implementation VoiceMessageTranscribeViewController

- (id)initWithAudiofileUrl:(NSURL *)audioFileUrl
{
    self = [super init];
    if (self) {
        self->_audioFileUrl = audioFileUrl;
    }
    
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];

    [NCAppBranding styleViewController:self];

    self.navigationItem.title = NSLocalizedString(@"Transcript", @"TRANSLATORS transcript of a voice-message");

    UIBarButtonItem *cancelButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel
                                                                                  target:self action:@selector(closeViewController)];
    
    self.navigationController.navigationBar.topItem.leftBarButtonItem = cancelButton;
    
    _activityIndicator = [[UIActivityIndicatorView alloc] init];
    if (@available(iOS 26.0, *)) {
        _activityIndicator.color = [UIColor labelColor];
    } else {
        _activityIndicator.color = [NCAppBranding themeTextColor];
    }
    
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:self->_activityIndicator];
    
    [_activityIndicator startAnimating];
    
    _supportedLocales = @[@"de", @"it", @"en", @"fr", @"es"];
    
    [self checkPermissionAndStartTranscription];
}

- (void)closeViewController
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)checkPermissionAndStartTranscription
{
    [SFSpeechRecognizer requestAuthorization:^(SFSpeechRecognizerAuthorizationStatus status){
        dispatch_async(dispatch_get_main_queue(), ^{
            if (status != SFSpeechRecognizerAuthorizationStatusAuthorized) {
                [self showSpeechRecognitionNotAvailable];
                return;
            }
            
            [self showLocaleSelection];
        });
    }];
}

- (void)showLocaleSelection
{
    UIAlertController *optionsActionSheet = [UIAlertController alertControllerWithTitle:nil
                                                                                message:nil
                                                                         preferredStyle:UIAlertControllerStyleActionSheet];
    
    // Use current locale for showing localized language names
    NSLocale *currentLocale = NSLocale.currentLocale;
    
    for (NSString *localeString in _supportedLocales) {
        NSLocale *speechLocale = [[NSLocale alloc] initWithLocaleIdentifier:localeString];
        SFSpeechRecognizer *speechRecognizer = [[SFSpeechRecognizer alloc] initWithLocale:speechLocale];
        
        if (!speechRecognizer.isAvailable || !speechRecognizer.supportsOnDeviceRecognition) {
            // We explicitly want to use on-device recognition
            continue;
        }
        
        UIAlertAction *localeAction = [UIAlertAction actionWithTitle:[currentLocale localizedStringForLanguageCode:localeString]
                                                                     style:UIAlertActionStyleDefault
                                                                   handler:^void (UIAlertAction *action) {
            [self transcribeWithLocale:speechLocale];
        }];
        
        [optionsActionSheet addAction:localeAction];
    }
    
    [optionsActionSheet addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", nil)
                                                           style:UIAlertActionStyleCancel
                                                         handler:^void (UIAlertAction *action) {
        [self closeViewController];
    }]];
    
    [self presentViewController:optionsActionSheet animated:YES completion:nil];
}

- (void)transcribeWithLocale:(NSLocale *)locale
{
    SFSpeechRecognizer *speechRecognizer = [[SFSpeechRecognizer alloc] initWithLocale:locale];
    SFSpeechURLRecognitionRequest *speechRecognitionRequest = [[SFSpeechURLRecognitionRequest alloc] initWithURL:_audioFileUrl];
    speechRecognitionRequest.requiresOnDeviceRecognition = YES;
    speechRecognitionRequest.shouldReportPartialResults = YES;
    
    [speechRecognizer recognitionTaskWithRequest:speechRecognitionRequest
                                   resultHandler:^(SFSpeechRecognitionResult * _Nullable result, NSError * _Nullable error)
    {
        if (error) {
            NSLog(@"Recognition task failed: %@", error.description);
            [self showSpeechRecognitionError:error.localizedDescription];
            return;
        }
        
        NSString *transcribedText = result.bestTranscription.formattedString;
        [self setTranscribedText:transcribedText isFinal:result.final];
    }];
    
}

- (void)showSpeechRecognitionError:(NSString *)errorDescription
{
    UIAlertController * alert = [UIAlertController
                                 alertControllerWithTitle:NSLocalizedString(@"Speech recognition failed", nil)
                                 message:errorDescription
                                 preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction* okButton = [UIAlertAction
                               actionWithTitle:NSLocalizedString(@"OK", nil)
                               style:UIAlertActionStyleDefault
                               handler:^(UIAlertAction * _Nonnull action) {
        [self closeViewController];
    }];
    
    [alert addAction:okButton];
    [self presentViewController:alert animated:YES completion:nil];
}

- (void)showSpeechRecognitionNotAvailable
{
    UIAlertController * alert = [UIAlertController
                                 alertControllerWithTitle:NSLocalizedString(@"Could not access speech recognition", nil)
                                 message:NSLocalizedString(@"Speech recognition access is not allowed. Check your settings.", nil)
                                 preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction* settingsButton = [UIAlertAction actionWithTitle:NSLocalizedString(@"Settings", nil)
                                                             style:UIAlertActionStyleDefault
                                                           handler:^(UIAlertAction * _Nonnull action) {
        [self closeViewController];
        [[UIApplication sharedApplication] openURL:[NSURL URLWithString:UIApplicationOpenSettingsURLString] options:@{} completionHandler:nil];
    }];
    [alert addAction:settingsButton];
    [alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"OK", nil)
                                              style:UIAlertActionStyleCancel
                                            handler:^(UIAlertAction * _Nonnull action) {
        [self closeViewController];
    }]];
    [self presentViewController:alert animated:YES completion:nil];
}

- (void)setTranscribedText:(NSString *)text isFinal:(BOOL)isFinal
{
    [self.transcribeTextView setText:text];
    
    if (isFinal) {
        [_activityIndicator stopAnimating];
        [_activityIndicator removeFromSuperview];
    }
}

@end

File: NextcloudTalk/Chat/VoiceMessageTranscribeViewController.swift
//
// SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
// SPDX-License-Identifier: GPL-3.0-or-later
//

import UIKit
import Speech

class VoiceMessageTranscribeViewController: UIViewController {

    @IBOutlet weak var transcribeTextView: UITextView!

    private let audioFileUrl: URL
    private let activityIndicator = UIActivityIndicatorView()
    private let supportedLocales = ["de", "it", "en", "fr", "es"]

    init(audiofileUrl audioFileUrl: URL) {
        self.audioFileUrl = audioFileUrl

        super.init(nibName: nil, bundle: nil)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        NCAppBranding.styleViewController(self)

        self.navigationItem.title = NSLocalizedString("Transcript", comment: "TRANSLATORS transcript of a voice-message")

        let cancelButton = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(closeViewController))
        self.navigationController?.navigationBar.topItem?.leftBarButtonItem = cancelButton

        if #available(iOS 26.0, *) {
            self.activityIndicator.color = .label
        } else {
            self.activityIndicator.color = NCAppBranding.themeTextColor()
        }

        self.navigationItem.rightBarButtonItem = UIBarButtonItem(customView: self.activityIndicator)

        self.activityIndicator.startAnimating()

        self.checkPermissionAndStartTranscription()
    }

    @objc func closeViewController() {
        self.dismiss(animated: true)
    }

    private func checkPermissionAndStartTranscription() {
        SFSpeechRecognizer.requestAuthorization { status in
            DispatchQueue.main.async { [weak self] in
                guard let self else { return }

                guard status == .authorized else {
                    self.showSpeechRecognitionNotAvailable()

                    return
                }

                self.showLocaleSelection()
            }
        }
    }

    private func showLocaleSelection() {
        let optionsActionSheet = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)

        // Use current locale for showing localized language names
        let currentLocale = Locale.current

        for localeString in self.supportedLocales {
            let speechLocale = Locale(identifier: localeString)
            let speechRecognizer = SFSpeechRecognizer(locale: speechLocale)

            // We explicitly want to use on-device recognition
            guard let speechRecognizer, speechRecognizer.isAvailable, speechRecognizer.supportsOnDeviceRecognition else { continue }

            optionsActionSheet.addAction(UIAlertAction(title: currentLocale.localizedString(forLanguageCode: localeString), style: .default) { [weak self] _ in
                self?.transcribe(with: speechLocale)
            })
        }

        optionsActionSheet.addAction(UIAlertAction(title: NSLocalizedString("Cancel", comment: ""), style: .cancel) { [weak self] _ in
            self?.closeViewController()
        })

        self.present(optionsActionSheet, animated: true)
    }

    private func transcribe(with locale: Locale) {
        let speechRecognizer = SFSpeechRecognizer(locale: locale)
        let speechRecognitionRequest = SFSpeechURLRecognitionRequest(url: self.audioFileUrl)
        speechRecognitionRequest.requiresOnDeviceRecognition = true
        speechRecognitionRequest.shouldReportPartialResults = true

        speechRecognizer?.recognitionTask(with: speechRecognitionRequest) { [weak self] result, error in
            guard let self else { return }

            if let error {
                NSLog("Recognition task failed: %@", (error as NSError).description)
                self.showSpeechRecognitionError(error.localizedDescription)

                return
            }

            self.setTranscribedText(result?.bestTranscription.formattedString, isFinal: result?.isFinal ?? false)
        }
    }

    private func showSpeechRecognitionError(_ errorDescription: String) {
        let alert = UIAlertController(title: NSLocalizedString("Speech recognition failed", comment: ""),
                                      message: errorDescription,
                                      preferredStyle: .alert)

        alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: ""), style: .default) { [weak self] _ in
            self?.closeViewController()
        })

        self.present(alert, animated: true)
    }

    private func showSpeechRecognitionNotAvailable() {
        let alert = UIAlertController(title: NSLocalizedString("Could not access speech recognition", comment: ""),
                                      message: NSLocalizedString("Speech recognition access is not allowed. Check your settings.", comment: ""),
                                      preferredStyle: .alert)

        alert.addAction(UIAlertAction(title: NSLocalizedString("Settings", comment: ""), style: .default) { [weak self] _ in
            self?.closeViewController()

            if let settingsUrl = URL(string: UIApplication.openSettingsURLString) {
                UIApplication.shared.open(settingsUrl)
            }
        })

        alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: ""), style: .cancel) { [weak self] _ in
            self?.closeViewController()
        })

        self.present(alert, animated: true)
    }

    private func setTranscribedText(_ text: String?, isFinal: Bool) {
        self.transcribeTextView.text = text

        if isFinal {
            self.activityIndicator.stopAnimating()
            self.activityIndicator.removeFromSuperview()
        }
    }
}
